DB_File.1
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH DB_FILE 1 "perl " "5/Sep/96" "User Contributed Perl Documentation"
.IX Title "DB_FILE 1"
.UC
.IX Name "DB_File - Perl5 access to Berkeley DB"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
DB_File \- Perl5 access to Berkeley DB
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.PP
.Vb 6
\& use DB_File ;
\& use strict 'untie' ;
\& 
\& [$X =] tie %hash,  'DB_File', [$filename, $flags, $mode, $DB_HASH] ;
\& [$X =] tie %hash,  'DB_File', $filename, $flags, $mode, $DB_BTREE ;
\& [$X =] tie @array, 'DB_File', $filename, $flags, $mode, $DB_RECNO ;
.Ve
.Vb 6
\& $status = $X->del($key [, $flags]) ;
\& $status = $X->put($key, $value [, $flags]) ;
\& $status = $X->get($key, $value [, $flags]) ;
\& $status = $X->seq($key, $value, $flags) ;
\& $status = $X->sync([$flags]) ;
\& $status = $X->fd ;
.Ve
.Vb 4
\& # BTREE only
\& $count = $X->get_dup($key) ;
\& @list  = $X->get_dup($key) ;
\& %list  = $X->get_dup($key, 1) ;
.Ve
.Vb 6
\& # RECNO only
\& $a = $X->length;
\& $a = $X->pop ;
\& $X->push(list);
\& $a = $X->shift;
\& $X->unshift(list);
.Ve
.Vb 2
\& untie %hash ;
\& untie @array ;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\fBDB_File\fR is a module which allows Perl programs to make use of the
facilities provided by Berkeley DB.  If you intend to use this
module you should really have a copy of the Berkeley DB manual pages at
hand. The interface defined here mirrors the Berkeley DB interface
closely.
.PP
Berkeley DB is a C library which provides a consistent interface to a
number of database formats.  \fBDB_File\fR provides an interface to all
three of the database types currently supported by Berkeley DB.
.PP
The file types are:
.Ip "\fB\s-1DB_HASH\s0\fR" 5
.IX Item "\fB\s-1DB_HASH\s0\fR"
This database type allows arbitrary key/value pairs to be stored in data
files. This is equivalent to the functionality provided by other
hashing packages like \s-1DBM\s0, \s-1NDBM\s0, \s-1ODBM\s0, \s-1GDBM\s0, and \s-1SDBM\s0. Remember though,
the files created using \s-1DB_HASH\s0 are not compatible with any of the
other packages mentioned.
.Sp
A default hashing algorithm, which will be adequate for most
applications, is built into Berkeley \s-1DB\s0. If you do need to use your own
hashing algorithm it is possible to write your own in Perl and have
\fBDB_File\fR use it instead.
.Ip "\fB\s-1DB_BTREE\s0\fR" 5
.IX Item "\fB\s-1DB_BTREE\s0\fR"
The btree format allows arbitrary key/value pairs to be stored in a
sorted, balanced binary tree.
.Sp
As with the \s-1DB_HASH\s0 format, it is possible to provide a user defined
Perl routine to perform the comparison of keys. By default, though, the
keys are stored in lexical order.
.Ip "\fB\s-1DB_RECNO\s0\fR" 5
.IX Item "\fB\s-1DB_RECNO\s0\fR"
\s-1DB_RECNO\s0 allows both fixed-length and variable-length flat text files
to be manipulated using the same key/value pair interface as in \s-1DB_HASH\s0
and \s-1DB_BTREE\s0.  In this case the key will consist of a record (line)
number.
.Sh "How does DB_File interface to Berkeley \s-1DB\s0?"
.IX Subsection "How does DB_File interface to Berkeley \s-1DB\s0?"
\fBDB_File\fR allows access to Berkeley \s-1DB\s0 files using the \fItie()\fR mechanism
in Perl 5 (for full details, see the \f(CWtie()\fR entry in the \fIperlfunc\fR manpage). This facility
allows \fBDB_File\fR to access Berkeley \s-1DB\s0 files using either an
associative array (for \s-1DB_HASH\s0 & \s-1DB_BTREE\s0 file types) or an ordinary
array (for the \s-1DB_RECNO\s0 file type).
.PP
In addition to the \fItie()\fR interface, it is also possible to access most
of the functions provided in the Berkeley \s-1DB\s0 \s-1API\s0 directly.
See the section on \fI\s-1THE\s0 \s-1API\s0 \s-1INTERFACE\s0\fR.
.Sh "Opening a Berkeley \s-1DB\s0 Database File"
.IX Subsection "Opening a Berkeley \s-1DB\s0 Database File"
Berkeley \s-1DB\s0 uses the function \fIdbopen()\fR to open or create a database.
Here is the C prototype for \fIdbopen()\fR:
.PP
.Vb 3
\&      DB*
\&      dbopen (const char * file, int flags, int mode, 
\&              DBTYPE type, const void * openinfo)
.Ve
The parameter \f(CWtype\fR is an enumeration which specifies which of the 3
interface methods (\s-1DB_HASH\s0, \s-1DB_BTREE\s0 or \s-1DB_RECNO\s0) is to be used.
Depending on which of these is actually chosen, the final parameter,
\fIopeninfo\fR points to a data structure which allows tailoring of the
specific interface method.
.PP
This interface is handled slightly differently in \fBDB_File\fR. Here is
an equivalent call using \fBDB_File\fR:
.PP
.Vb 1
\&        tie %array, 'DB_File', $filename, $flags, $mode, $DB_HASH ;
.Ve
The \f(CWfilename\fR, \f(CWflags\fR and \f(CWmode\fR parameters are the direct
equivalent of their \fIdbopen()\fR counterparts. The final parameter \f(CW$DB_HASH\fR
performs the function of both the \f(CWtype\fR and \f(CWopeninfo\fR parameters in
\fIdbopen()\fR.
.PP
In the example above \f(CW$DB_HASH\fR is actually a pre-defined reference to a
hash object. \fBDB_File\fR has three of these pre-defined references.
Apart from \f(CW$DB_HASH\fR, there is also \f(CW$DB_BTREE\fR and \f(CW$DB_RECNO\fR.
.PP
The keys allowed in each of these pre-defined references is limited to
the names used in the equivalent C structure. So, for example, the
\f(CW$DB_HASH\fR reference will only allow keys called \f(CWbsize\fR, \f(CWcachesize\fR,
\f(CWffactor\fR, \f(CWhash\fR, \f(CWlorder\fR and \f(CWnelem\fR. 
.PP
To change one of these elements, just assign to it like this:
.PP
.Vb 1
\&        $DB_HASH->{'cachesize'} = 10000 ;
.Ve

Changes

1.860 1 February 2026

* Add automation to the workflows to detect and use all the available versions of perl  available.
  Sun Feb 1 16:12:39 2026 +0000
  b7ccaa172ae2d844b7bae7a9829bcaee463ad186

* Update to version 1.860
  Sun Feb 1 16:05:15 2026 +0000
  9f759912ce94924c64c933f9586fc6fb50eb4b35

* Update copyright date to 2026
  Thu Jan 1 15:33:59 2026 +0000
  8fac218ef6f153294ab3bdba2742210950c751a7

* Bump actions/cache from 4 to 5
  Thu Jan 1 15:30:37 2026 +0000
  8ee97cff1e0d237682558b8be3e2fbfeb385ac1f

* Bump actions/checkout from 5 to 6
  Mon Dec 1 14:20:34 2025 +0000
  ebdcc8e2a307ab5ad1b0487c93fa41e7fa643ee0

* Add dependabot.yml to police workflow files
  Thu Sep 4 17:33:28 2025 +0100
  ba0fc74444a846a76eb670daf1b87cfa0ac2f215

* Fix links to dbopen manpage (#17)
  Thu Apr 10 11:53:24 2025 +0100
  06ea397e5d654ebeb9d22a6993c8aaebf710ce9b


1.859 21 August 2023

   * silence clang compiler warnings
     https://github.com/pmqs/DB_File/issues/13
     Sun Aug 20 16:30:05 2023 +0100
     6f8902efe7b2782ad418ea53ddaad08ed9bebf9b

1.858 13 May 2022

   * Update to license wording
      Fri May 13 07:50:35 2022 +0100
      5089ede9dbc01ea9e47f5cb96bf9c8706ad1794d

1.857 25 February 2022

   * Silence gcc12 'empty-body' warning
     https://github.com/pmqs/DB_File/issues/10
     22c4850b70d0c32eb4acb9c987868274a61072ec

1.856 22 June 2021

   * Ensure that all *.bak files are cleaned up
     https://github.com/pmqs/DB_File/pull/9
     96bcf494625897ab4fcda322481dd6f730bdb668


1.855 13 October 2020

   * Stop shipping DB_File_BS for the NeXT operating system.
     https://github.com/pmqs/DB_File/pull/7
     e5f329e066bf99f683c895f4a8b8ea328804bbb1
     80e0bb7b655a2cb035292b3ae4b095d569314bdf

1.854 16 September 2020

   * Prefer direct notation over indirect (#4)
     fae7c3898579fab1a45b52d2c1d9c19370b21c84

   * Make hint/BS snippets strict compliant.
     fb503ef27fae2cdce84d1c341e4991917da3bea6

   * trim whitespace
     6d8428d6f038992e3abbc5a2c7cf17b6d4a75382

1.853 5 January 2020

   * Memory leak if ParseOpenInfo calls croak_and_free
     https://github.com/pmqs/DB_File/issues/3
     9e2c8d6c3a35acb71358f440c93defa3d8339420

   * Add Address Sanatizer suppressions file
     a6a26f4878d6be13aad69a3f5b6019e7482a8992

   * Add prereq test
     a631884b3ca11919142c6be8a64e668730b83469

   * Documemtation updates
     4c53b38531b547ae293967c852ff21412eb6d840

   * clang warning in ppport.h
     update to latest ppport.h
     https://github.com/pmqs/DB_File/issues/2
     36d6ae54edfc1df872f5e66c93bda05cbfefefa7

   * #125853 - RT links in Changes file are wrong
     Changes entry for 1.842 has both CPAN & Perl RT links.
     Use the full URL to avoid ambiguity.
     f06a9235373747d0f5c6a95caf504174ffb19c44

1.852 21 April 2019

   * ParseOpenInfo() leaks if it croaks
     DB_File_type memory not freed when ParseOpenInfo croaks.
     https://github.com/pmqs/DB_File/issues/1

   * Fix whitespace in DB_File.xs

1.851 4 April Feb 2019

   * Fix typo in MANIFEST with .travis.yml file

1.850 10 Feb 2019

   * Added source to github
     https://github.com/pmqs/DB_File

   * Added .travis.yml

   * Added t/meta-json.t & t/meta-yaml.t

1.843 2 November 2018

   * hints file (sourced from
     http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/lang/perl5/files/)
     Highlighed by Sevan Janiyan <venture37@geeklan.co.uk>

        hints/bitrig.pl
        hints/netbsd.pl
        hints/minix.pl
        hints/openbsd.pl

1.842 14 July 2018

   * Document the MaocOS build failures & how MacPort can work around the issue.
     https://rt.cpan.org/Public/Bug/Display.html?id=125238
     https://rt.cpan.org/Public/Bug/Display.html?id=125585
     https://rt.perl.org/Public/Bug/Display.html?id=133280

   * Write all test db files into a directory created by File::Temp
     Protects against any possible races if running a parallel build.

   * #96357: DB_File destructor is not thread-safe
     Added CLONE_SKIP to DB_File.pm and included the test db-threads.t from the patch.
     Note - This only works with perl >= 5.8.7

1.841 2 Apr 2018

   * #124944 allow ppport.h-less builds in core

1.840 29 Dec 2016

   * #119569: Failed to install DB_File-1.839 into Linux

1.839 29 Dec 2016

   * #119557: DB_File.xs: unused-parameter warnings
   * #107642: unused arg

1.838 15 May 2016

   * #101342: R_SETCURSOR compiler warning
     #111726: Coverity finding: flagSet() test always false

1.837 9 Jan 2016

   * DB_File: database UTF-8 documentation
     RT #110248


blib/DB_File.pm
# DB_File.pm -- Perl 5 interface to Berkeley DB 
#
# written by Paul Marquess (pmarquess@bfsec.bt.co.uk)
# last modified 7th October 1995
# version 1.0

package DB_File::HASHINFO ;
use Carp;

sub TIEHASH
{
    bless {} ;
}

%elements = ( 'bsize'     => 0,
              'ffactor'   => 0,
              'nelem'     => 0,
              'cachesize' => 0,
              'hash'      => 0,
              'lorder'    => 0
            ) ;

sub FETCH 
{  
    return $_[0]{$_[1]} if defined $elements{$_[1]}  ;

    croak "DB_File::HASHINFO::FETCH - Unknown element '$_[1]'" ;
}


sub STORE 
{
    if ( defined $elements{$_[1]} )
    {
        $_[0]{$_[1]} = $_[2] ;
        return ;
    }
    
    croak "DB_File::HASHINFO::STORE - Unknown element '$_[1]'" ;
}

sub DELETE 
{
    if ( defined $elements{$_[1]} )
    {
        delete ${$_[0]}{$_[1]} ;
        return ;
    }
    
    croak "DB_File::HASHINFO::DELETE - Unknown element '$_[1]'" ;
}


sub DESTROY {undef %{$_[0]} }
sub FIRSTKEY { croak "DB_File::HASHINFO::FIRSTKEY is not implemented" }
sub NEXTKEY { croak "DB_File::HASHINFO::NEXTKEY is not implemented" }
sub EXISTS { croak "DB_File::HASHINFO::EXISTS is not implemented" }
sub CLEAR { croak "DB_File::HASHINFO::CLEAR is not implemented" }

package DB_File::BTREEINFO ;
use Carp;

sub TIEHASH
{
    bless {} ;
}

%elements = ( 'flags'	=> 0,
              'cachesize'  => 0,
              'maxkeypage' => 0,
              'minkeypage' => 0,
              'psize'      => 0,
              'compare'    => 0,
              'prefix'     => 0,
              'lorder'     => 0
            ) ;

sub FETCH 
{  
    return $_[0]{$_[1]} if defined $elements{$_[1]}  ;

    croak "DB_File::BTREEINFO::FETCH - Unknown element '$_[1]'" ;
}


sub STORE 
{
    if ( defined $elements{$_[1]} )
    {
        $_[0]{$_[1]} = $_[2] ;
        return ;
    }
    
    croak "DB_File::BTREEINFO::STORE - Unknown element '$_[1]'" ;
}

sub DELETE 
{
    if ( defined $elements{$_[1]} )
    {
        delete ${$_[0]}{$_[1]} ;
        return ;
    }
    
    croak "DB_File::BTREEINFO::DELETE - Unknown element '$_[1]'" ;
}


sub DESTROY {undef %{$_[0]} }
sub FIRSTKEY { croak "DB_File::BTREEINFO::FIRSTKEY is not implemented" }
sub NEXTKEY { croak "DB_File::BTREEINFO::NEXTKEY is not implemented" }
sub EXISTS { croak "DB_File::BTREEINFO::EXISTS is not implemented" }
sub CLEAR { croak "DB_File::BTREEINFO::CLEAR is not implemented" }

package DB_File::RECNOINFO ;
use Carp;

sub TIEHASH
{
    bless {} ;
}

%elements = ( 'bval'      => 0,
              'cachesize' => 0,
              'psize'     => 0,
              'flags'     => 0,
              'lorder'    => 0,
              'reclen'    => 0,
              'bfname'    => 0
            ) ;
sub FETCH 
{  
    return $_[0]{$_[1]} if defined $elements{$_[1]}  ;

    croak "DB_File::RECNOINFO::FETCH - Unknown element '$_[1]'" ;
}


sub STORE 
{
    if ( defined $elements{$_[1]} )
    {
        $_[0]{$_[1]} = $_[2] ;
        return ;
    }
    
    croak "DB_File::RECNOINFO::STORE - Unknown element '$_[1]'" ;
}

sub DELETE 
{
    if ( defined $elements{$_[1]} )
    {
        delete ${$_[0]}{$_[1]} ;
        return ;
    }
    
    croak "DB_File::RECNOINFO::DELETE - Unknown element '$_[1]'" ;
}


sub DESTROY {undef %{$_[0]} }
sub FIRSTKEY { croak "DB_File::RECNOINFO::FIRSTKEY is not implemented" }
sub NEXTKEY { croak "DB_File::RECNOINFO::NEXTKEY is not implemented" }
sub EXISTS { croak "DB_File::BTREEINFO::EXISTS is not implemented" }
sub CLEAR { croak "DB_File::BTREEINFO::CLEAR is not implemented" }



package DB_File ;
use Carp;

$VERSION = 1.0 ;

#typedef enum { DB_BTREE, DB_HASH, DB_RECNO } DBTYPE;
$DB_BTREE = TIEHASH DB_File::BTREEINFO ;
$DB_HASH  = TIEHASH DB_File::HASHINFO ;
$DB_RECNO = TIEHASH DB_File::RECNOINFO ;

require TieHash;
require Exporter;
use AutoLoader;
require DynaLoader;
@ISA = qw(TieHash Exporter DynaLoader);
@EXPORT = qw(
        $DB_BTREE $DB_HASH $DB_RECNO 
	BTREEMAGIC
	BTREEVERSION
	DB_LOCK
	DB_SHMEM
	DB_TXN
	HASHMAGIC
	HASHVERSION
	MAX_PAGE_NUMBER
	MAX_PAGE_OFFSET
	MAX_REC_NUMBER
	RET_ERROR
	RET_SPECIAL
	RET_SUCCESS
	R_CURSOR
	R_DUP
	R_FIRST
	R_FIXEDLEN
	R_IAFTER
	R_IBEFORE
	R_LAST
	R_NEXT
	R_NOKEY
	R_NOOVERWRITE
	R_PREV
	R_RECNOSYNC
	R_SETCURSOR
	R_SNAPSHOT
	__R_UNUSED
);

sub AUTOLOAD {
    local($constname);
    ($constname = $AUTOLOAD) =~ s/.*:://;
    $val = constant($constname, @_ ? $_[0] : 0);
    if ($! != 0) {
	if ($! =~ /Invalid/) {
	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
	    goto &AutoLoader::AUTOLOAD;
	}
	else {
	    ($pack,$file,$line) = caller;
	    croak "Your vendor has not defined DB macro $constname, used at $file line $line.
";
	}
    }
    eval "sub $AUTOLOAD { $val }";
    goto &$AUTOLOAD;
}

@liblist = ();
@liblist = split ' ', $Config::Config{"DB_File_loadlibs"} 
    if defined $Config::Config{"DB_File_loadlibs"};

bootstrap DB_File @liblist;

# Preloaded methods go here.  Autoload methods go after __END__, and are
# processed by the autosplit program.

1;
__END__

=cut

=head1 NAME

DB_File - Perl5 access to Berkeley DB

=head1 SYNOPSIS

 use DB_File ;
  
 [$X =] tie %hash,  DB_File, $filename [, $flags, $mode, $DB_HASH] ;
 [$X =] tie %hash,  DB_File, $filename, $flags, $mode, $DB_BTREE ;
 [$X =] tie @array, DB_File, $filename, $flags, $mode, $DB_RECNO ;
   
 $status = $X->del($key [, $flags]) ;
 $status = $X->put($key, $value [, $flags]) ;
 $status = $X->get($key, $value [, $flags]) ;
 $status = $X->seq($key, $value [, $flags]) ;
 $status = $X->sync([$flags]) ;
 $status = $X->fd ;
    
 untie %hash ;
 untie @array ;

=head1 DESCRIPTION

B<DB_File> is a module which allows Perl programs to make use of the
facilities provided by Berkeley DB.  If you intend to use this
module you should really have a copy of the Berkeley DB manualpage at
hand. The interface defined here mirrors the Berkeley DB interface
closely.

Berkeley DB is a C library which provides a consistent interface to a
number of database formats.  B<DB_File> provides an interface to all
three of the database types currently supported by Berkeley DB.

The file types are:

=over 5

=item DB_HASH

This database type allows arbitrary key/data pairs to be stored in data
files. This is equivalent to the functionality provided by other
hashing packages like DBM, NDBM, ODBM, GDBM, and SDBM. Remember though,
the files created using DB_HASH are not compatible with any of the
other packages mentioned.

A default hashing algorithm, which will be adequate for most
applications, is built into Berkeley DB. If you do need to use your own
hashing algorithm it is possible to write your own in Perl and have
B<DB_File> use it instead.

=item DB_BTREE

The btree format allows arbitrary key/data pairs to be stored in a
sorted, balanced binary tree.

As with the DB_HASH format, it is possible to provide a user defined
Perl routine to perform the comparison of keys. By default, though, the
keys are stored in lexical order.

=item DB_RECNO

DB_RECNO allows both fixed-length and variable-length flat text files
to be manipulated using the same key/value pair interface as in DB_HASH
and DB_BTREE.  In this case the key will consist of a record (line)
number.

=back

=head2 How does DB_File interface to Berkeley DB?

B<DB_File> allows access to Berkeley DB files using the tie() mechanism
in Perl 5 (for full details, see L<perlfunc/tie()>). This facility
allows B<DB_File> to access Berkeley DB files using either an
associative array (for DB_HASH & DB_BTREE file types) or an ordinary
array (for the DB_RECNO file type).

In addition to the tie() interface, it is also possible to use most of
the functions provided in the Berkeley DB API.

=head2 Differences with Berkeley DB

Berkeley DB uses the function dbopen() to open or create a database.
Below is the C prototype for dbopen().

      DB*
      dbopen (const char * file, int flags, int mode, 
              DBTYPE type, const void * openinfo)

The parameter C<type> is an enumeration which specifies which of the 3
interface methods (DB_HASH, DB_BTREE or DB_RECNO) is to be used.
Depending on which of these is actually chosen, the final parameter,
I<openinfo> points to a data structure which allows tailoring of the
specific interface method.

This interface is handled slightly differently in B<DB_File>. Here is
an equivalent call using B<DB_File>.

        tie %array, DB_File, $filename, $flags, $mode, $DB_HASH ;

The C<filename>, C<flags> and C<mode> parameters are the direct
equivalent of their dbopen() counterparts. The final parameter $DB_HASH
performs the function of both the C<type> and C<openinfo> parameters in
dbopen().

In the example above $DB_HASH is actually a reference to a hash
object. B<DB_File> has three of these pre-defined references. Apart
from $DB_HASH, there is also $DB_BTREE and $DB_RECNO.

The keys allowed in each of these pre-defined references is limited to
the names used in the equivalent C structure. So, for example, the
$DB_HASH reference will only allow keys called C<bsize>, C<cachesize>,
C<ffactor>, C<hash>, C<lorder> and C<nelem>.

To change one of these elements, just assign to it like this

	$DB_HASH{cachesize} = 10000 ;


=head2 RECNO


In order to make RECNO more compatible with Perl the array offset for all
RECNO arrays begins at 0 rather than 1 as in Berkeley DB.


=head2 In Memory Databases

Berkeley DB allows the creation of in-memory databases by using NULL
(that is, a C<(char *)0 in C) in place of the filename.  B<DB_File>
uses C<undef> instead of NULL to provide this functionality.


=head2 Using the Berkeley DB Interface Directly

As well as accessing Berkeley DB using a tied hash or array, it is also
possible to make direct use of most of the functions defined in the
Berkeley DB documentation.


To do this you need to remember the return value from the tie.

	$db = tie %hash, DB_File, "filename"

Once you have done that, you can access the Berkeley DB API functions
directly.

	$db->put($key, $value, R_NOOVERWRITE) ;

All the functions defined in L<dbx(3X)> are available except for
close() and dbopen() itself. The B<DB_File> interface to these
functions have been implemented to mirror the the way Berkeley DB
works. In particular note that all the functions return only a status
value. Whenever a Berkeley DB function returns data via one of its
parameters, the B<DB_File> equivalent does exactly the same.

All the constants defined in L<dbopen> are also available.

Below is a list of the functions available.

=over 5

=item get

Same as in C<recno> except that the flags parameter is optional.
Remember the value associated with the key you request is returned in
the $value parameter.

=item put

As usual the flags parameter is optional. 

If you use either the R_IAFTER or R_IBEFORE flags, the key parameter
will have the record number of the inserted key/value pair set.

=item del

The flags parameter is optional.

=item fd

As in I<recno>.

=item seq

The flags parameter is optional.

Both the key and value parameters will be set.

=item sync

The flags parameter is optional.

=back

=head1 EXAMPLES

It is always a lot easier to understand something when you see a real
example. So here are a few.

=head2 Using HASH

	use DB_File ;
	use Fcntl ;
	
	tie %h,  "DB_File", "hashed", O_RDWR|O_CREAT, 0640, $DB_HASH ;
	
	# Add a key/value pair to the file
	$h{"apple"} = "orange" ;
	
	# Check for existence of a key
	print "Exists\n" if $h{"banana"} ;
	
	# Delete 
	delete $h{"apple"} ;
	
	untie %h ;

=head2 Using BTREE

Here is sample of code which used BTREE. Just to make life more
interesting the default comparision function will not be used. Instead
a Perl sub, C<Compare()>, will be used to do a case insensitive
comparison.

        use DB_File ;
        use Fcntl ;
	 
	sub Compare
        {
	    my ($key1, $key2) = @_ ;
	
	    "\L$key1" cmp "\L$key2" ;
	}
	
        $DB_BTREE->{compare} = 'Compare' ;
	 
        tie %h,  'DB_File', "tree", O_RDWR|O_CREAT, 0640, $DB_BTREE ;
	 
        # Add a key/value pair to the file
        $h{'Wall'} = 'Larry' ;
        $h{'Smith'} = 'John' ;
	$h{'mouse'} = 'mickey' ;
	$h{'duck'}   = 'donald' ;
	 
        # Delete
        delete $h{"duck"} ;
	 
	# Cycle through the keys printing them in order.
	# Note it is not necessary to sort the keys as
	# the btree will have kept them in order automatically.
	foreach (keys %h)
	  { print "$_\n" }
	
        untie %h ;

Here is the output from the code above.

	mouse
	Smith
	Wall


=head2 Using RECNO

	use DB_File ;
	use Fcntl ;
	
	$DB_RECNO->{psize} = 3000 ;
	
	tie @h,  DB_File, "text", O_RDWR|O_CREAT, 0640, $DB_RECNO ;
	
	# Add a key/value pair to the file
	$h[0] = "orange" ;
	
	# Check for existence of a key
	print "Exists\n" if $h[1] ;
	
	untie @h ;



=head1 CHANGES

=head2 0.1

First Release.

=head2 0.2

When B<DB_File> is opening a database file it no longer terminates the
process if I<dbopen> returned an error. This allows file protection
errors to be caught at run time. Thanks to Judith Grass
<grass@cybercash.com> for spotting the bug.

=head2 0.3

Added prototype support for multiple btree compare callbacks.

=head2 1.0

B<DB_File> has been in use for over a year. To reflect that, the
version number has been incremented to 1.0.

Added complete support for multiple concurrent callbacks.

Using the I<push> method on an empty list didn't work properly. This
has been fixed.

=head2 1.001

Fixed a core dump problem with SunOS.

=head1 WARNINGS

If you happen find any other functions defined in the source for this
module that have not been mentioned in this document -- beware.  I may
drop them at a moments notice.

If you cannot find any, then either you didn't look very hard or the
moment has passed and I have dropped them.

=head1 BUGS

Some older versions of Berkeley DB had problems with fixed length
records using the RECNO file format. The newest version at the time of
writing was 1.85 - this seems to have fixed the problems with RECNO.

I am sure there are bugs in the code. If you do find any, or can
suggest any enhancements, I would welcome your comments.

=head1 AVAILABILITY

Berkeley DB is available via the hold C<ftp.cs.berkeley.edu> in the
directory C</ucb/4bsd/db.tar.gz>.  It is I<not> under the GPL.

=head1 SEE ALSO

L<perl(1)>, L<dbopen(3)>, L<hash(3)>, L<recno(3)>, L<btree(3)> 

Berkeley DB is available from F<ftp.cs.berkeley.edu> in the directory
F</ucb/4bsd>.

=head1 AUTHOR

The DB_File interface was written by Paul Marquess
<pmarquess@bfsec.bt.co.uk>.
Questions about the DB system itself may be addressed to Keith Bostic
<bostic@cs.berkeley.edu>.

=cut
